docs/design/task-launch-order-within-job.md

```diff
@@ -0,0 +1,79 @@
+# Task Launch Order Within Job
+
+@[hwdef](https://github.com/hwdef), @[Thor-wl](https://github.com/Thor-wl); July 19th, 2021
+
+## Introduction
+
+This feature provides the ability to customize the order in which tasks are launched.
+
+## Scope
+
+### In Scope
+
+* Reasons why this feature is needed
+* Define the API
+
+### Out of Scope
+
+* Start order of job
+* Task start sequence between multiple jobs
+* Dependency completion state of the task start sequence
+
+## Scenarios
+
+* MPI Job. The worker pods need to be started first and then the master pod can run. Elsewise, the master pod can't setup the ssh tunnel and thus failed, this will add unnecessary waste of resources. In this case, MPI worker pods need to be in the running state, and then the master pod can start. Very similary, for the TensorFlow job, the TF master pod must be started first and then the TF worker pods.
+
+## Scene Comparison
+
+| example   | number of tasks | task dependencies | concurrent execution | Solution Status   | Disadvantages of the current solution |
+| ---------- | -------- | ------------ | -------- | -------------------------------------------------- | -------------------------- |
+| MPI        | 2        | linear dependencies | yes    | 1. Add initcontainer to the task<br/>2.Using user-written check scripts in initcontainer | 1.Increase the cost of use for users<br />2.Still using resources while waiting |
+| matlab     | 2        | linear dependencies | yes    | 1. Add initcontainer to the task<br/>2.Using user-written check scripts in initcontainer |1.Increase the cost of use for users<br />2.Still using resources while waiting|
+| tensorflow | n>=2   | linear dependencies | yes    | 1. Add initcontainer to the task<br/>2.Using user-written check scripts in initcontainer |1.Increase the cost of use for users<br />2.Still using resources while waiting|
+
+## Requirement
+
+Based on the scenarios listed above, the dependencies can be abstracted as:
+
+* Task-A depends on task-B, which means A must be started first and then B.
+* Triggering policy, in our cases, there may be only one trigger policy, which is the running state
+
+For the ease of end user's experince, we need to unify the way of composing a complicated jobs with lots of tasks, instead of letting user handle the complexcity themselvs using init-containers or other workflow tools. So we need to  have an more advanced VCJob that has below abilities:
+
+## Design
+
+### Field
+
+Add a field in `vcjob.spec.task.DependentTask`, it represents the name of the task that this task needs to depend on
```

Comment:
let's add an object instead of a field which is more extendable :)

Reply:
@k82cn 
After yesterday’s discussion, I think that several other reviewers do not want the possibility of expansion of this feature.Currently only the running state is retained. If a field has only one value, then the field should not exist.

```
dependency:
   name: taskA
   status: running
```
For example, in an example like the one above, there is no reason for the status field to exist

Reply:
> After yesterday’s discussion, I think that several other reviewers do not want the possibility of expansion of this feature.

Extensibility is important! It's ok to support only one task right now, but it's better to be extendable, something like that:

```
depends:
   name: taskA
```

Reply:
1. we should support multi tasks.
2. In yesterday's meeting, I felt that other reviewers did not pay attention to scalability. Please unify internally
3. `depends` just have one field named `name`, this kind of design is an over-design.Yesterday, we made it clear that we will only solve existing problems and not consider future scenarios.